src/kernel/keyboard.zig

```diff
@@ -0,0 +1,353 @@
+        const key = self.queue[self.queue_front];
+        _ = @addWithOverflow(QueueIndex, self.queue_front, 1, &self.queue_front);
+        return key;
+    }
+
+    test "init" {
+        const keyboard = Keyboard.init();
+        testing.expectEqual(keyboard.queue_front, 0);
+        testing.expectEqual(keyboard.queue_end, 0);
+    }
+
+    test "isEmpty" {
+        var keyboard = Keyboard.init();
+        testing.expect(keyboard.isEmpty());
+
+        keyboard.queue_end += 1;
+        testing.expect(!keyboard.isEmpty());
+
+        keyboard.queue_front += 1;
+        testing.expect(keyboard.isEmpty());
+
+        keyboard.queue_end = std.math.maxInt(QueueIndex);
+        keyboard.queue_front = 0;
+        testing.expect(!keyboard.isEmpty());
+
+        keyboard.queue_front = std.math.maxInt(QueueIndex);
+        testing.expect(keyboard.isEmpty());
+    }
+
+    test "isFull" {
+        var keyboard = Keyboard.init();
+        testing.expect(!keyboard.isFull());
+
+        keyboard.queue_end += 1;
+        testing.expect(!keyboard.isFull());
+
+        keyboard.queue_front += 1;
+        testing.expect(!keyboard.isFull());
+
+        keyboard.queue_end = 0;
+        testing.expect(keyboard.isFull());
+
+        keyboard.queue_front = 0;
+        keyboard.queue_end = std.math.maxInt(QueueIndex);
+        testing.expect(keyboard.isFull());
+    }
+
+    test "writeKey" {
+        var keyboard = Keyboard.init();
+
+        comptime var i = 0;
+        inline while (i < QUEUE_SIZE - 1) : (i += 1) {
+            testing.expectEqual(keyboard.writeKey(.{
+                .position = @intToEnum(KeyPosition, i),
+                .released = false,
+            }), true);
+            testing.expectEqual(keyboard.queue[i].position, @intToEnum(KeyPosition, i));
+            testing.expectEqual(keyboard.queue_end, i + 1);
+            testing.expectEqual(keyboard.queue_front, 0);
+        }
+
+        testing.expectEqual(keyboard.writeKey(.{
+            .position = @intToEnum(KeyPosition, 33),
+            .released = false,
+        }), false);
+        testing.expect(keyboard.isFull());
+    }
+
+    test "readKey" {
+        var keyboard = Keyboard.init();
+
+        comptime var i = 0;
+        inline while (i < QUEUE_SIZE - 1) : (i += 1) {
+            testing.expectEqual(keyboard.writeKey(.{
+                .position = @intToEnum(KeyPosition, i),
+                .released = false,
+            }), true);
+        }
+
+        i = 0;
+        inline while (i < QUEUE_SIZE - 1) : (i += 1) {
+            testing.expectEqual(keyboard.readKey().?.position, @intToEnum(KeyPosition, i));
+            testing.expectEqual(keyboard.queue_end, QUEUE_SIZE - 1);
+            testing.expectEqual(keyboard.queue_front, i + 1);
+        }
+
+        testing.expect(keyboard.isEmpty());
+        testing.expectEqual(keyboard.readKey(), null);
+    }
+};
+
+/// The registered keyboards
+var keyboards: ArrayList(*Keyboard) = undefined;
```

Comment:
Are we expecting to have more than one keyboard?

Reply:
I think it would be good to support it. Modern OSs do AFAIK.

Reply:
We should